File: src/main/java/com/example/chatService/component/WebSocketDisconnectListener.java
package com.example.chatService.component;

import com.example.chatService.dto.UserRoomSession;
import com.example.chatService.redis.UserSessionRegistry;
import com.example.chatService.service.RoomParticipantService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.event.EventListener;
import org.springframework.messaging.simp.stomp.StompHeaderAccessor;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.messaging.SessionDisconnectEvent;

@Component
@RequiredArgsConstructor
@Slf4j
public class WebSocketDisconnectListener {

    private final RoomParticipantService roomParticipantService;
    private final UserSessionRegistry userSessionRegistry;

    @EventListener
    public void handleDisconnect(SessionDisconnectEvent event) {
        StompHeaderAccessor accessor = StompHeaderAccessor.wrap(event.getMessage());

        String sessionId = accessor.getSessionId();

        UserRoomSession info = userSessionRegistry.remove(sessionId);
        if (info == null) return;

        try {
            roomParticipantService.leaveRoom(info.getRoomId(), info.getUserId());
        } catch (Exception e) {
            log.warn("disconnect leave failed roomId={}, userId={}", info.getRoomId(), info.getUserId(), e);
        }
    }
}

File: src/main/java/com/example/chatService/component/WebSocketDisconnectListener.kt
package com.example.chatService.component

import com.example.chatService.redis.UserSessionRegistry
import com.example.chatService.service.RoomParticipantService
import org.slf4j.LoggerFactory
import org.springframework.context.event.EventListener
import org.springframework.messaging.simp.stomp.StompHeaderAccessor
import org.springframework.stereotype.Component
import org.springframework.web.socket.messaging.SessionDisconnectEvent

@Component
class WebSocketDisconnectListener(
    private val roomParticipantService: RoomParticipantService,
    private val userSessionRegistry: UserSessionRegistry,
) {
    @EventListener
    fun handleDisconnect(event: SessionDisconnectEvent) {
        val accessor = StompHeaderAccessor.wrap(event.message)
        val sessionId = accessor.sessionId ?: return
        val info = userSessionRegistry.remove(sessionId) ?: return

        try {
            roomParticipantService.leaveRoom(info.roomId, info.userId)
        } catch (e: Exception) {
            log.warn("disconnect leave failed roomId={}, userId={}", info.roomId, info.userId, e)
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(WebSocketDisconnectListener::class.java)
    }
}
